zcash_client_sqlite/src/wallet.rs

```diff
@@ -1964,6 +1974,77 @@ pub(crate) fn store_transaction_to_be_sent<P: consensus::Parameters>(
         }
     }
 
+    // Add the transaction to the set to be queried for transaction status. This is only necessary
+    // at present for fully transparent transactions, because any transaction with a shielded
+    // component will be detected via ordinary chain scanning and/or nullifier checking.
+    if !detectable_via_scanning {
+        queue_tx_retrieval(
+            wdb.conn.0,
+            std::iter::once(sent_tx.tx().txid()),
+            TxQueryType::Status,
+            None,
+        )?;
+    }
+
+    Ok(())
+}
+
+pub(crate) fn set_transaction_status(
+    conn: &rusqlite::Transaction,
+    txid: TxId,
+    status: TransactionStatus,
+) -> Result<(), SqliteClientError> {
+    // In order to have made a Status request, we must already have had the
+    // raw transaction (the only callers of `queue_tx_retrieval` are in contexts
+    // where we must have already called `put_tx_data`). Therefore, it is safe
+    // to unconditionally delete the request from `tx_retrieval_queue` below
+    // (both in the expired case and the case where it has been mined), because
+    // we already have all the data we need about this transaction.
```

Comment:
This is true for Status requests, but not true for Enhancement requests which subsume Status.

Reply:
Yes. The first commit of https://github.com/nuttycom/librustzcash/pull/39 fixes that.

Reply:
It occurs to me that https://github.com/zcash/librustzcash/pull/1473/files#diff-a1724f7547ecabc613a07520eaf7917d9084d4277219d186698b0c8e352495a4R2425-R2427 is superfluous; we should never have a `Status` request for a transaction unless we have the raw data. So one way to fix this would be to remove the query type argument from the `queue_tx_retrieval` function, and to set it based upon the presence or absence of raw transaction data.

Reply:
The comment change doesn't resolve my concern. `GetStatus` is indeed only generated when we have the raw transaction (as I noted), but `Enhancement` is generated by definition when we don't have the raw transaction, and its documentation states that when the caller is unable to look up the transaction, it must call this method.

The reason it is okay to delete for `Enhancement` requests is that:
- If a transaction can't be enhanced, we _assume_ that the caller calls this method with `TransactionStatus::TxidNotRecognized` (as the other two statuses mean that the chain data source knew about the transaction, and therefore must have observed it and could return its raw data), so we don't need to consider what happens in the other statuses.
- We only delete for `TransactionStatus::TxidNotRecognized` if the transaction is expired, at which point we don't _necessarily_ care about enhancement (though the user might still care about the memo), and it is very likely that we will be unable to ever enhance (as the tx will have been dropped from all mempools).

Reply:
Simplified in https://github.com/zcash/librustzcash/pull/1473/commits/5a1645ac76199a571a5a9270e12114e553f019e1

Reply:
It is always okay to delete _any_ sort of status request if a transaction is either (a) expired unmined, or (b) mined, in which case we know that we already have the transaction data, because we cannot generate a `GetStatus` request otherwise.

Reply:
I don't agree that 5a1645ac76199a571a5a9270e12114e553f019e1 is a simplification. It seems less explicit and requires more complicated reasoning about the sequence of states. The fact that we want to get the raw transaction when we don't have it and the status otherwise isn't fundamental, it's an implementation detail. We could have wanted to get both the status *and* the raw transaction in the case where we don't have the latter (which is what I originally thought we were doing until I looked closely at `store_decrypted_tx`).

Reply:
The fact is that an `Enhancement` request _is_ a `Status` request. That's what it means for an `Enhancement` request to subsume a `Status` request.

Reply:
@daira `Enhancement` _is_ getting the status; it subsumes / is a superset of `GetStatus`. That's why I raised my initial concern that this method was being written ostensibly assuming it would only be called via `GetStatus`, which was false.

As @nuttycom says, the actual implementation does correctly work with both `GetStatus` and `Enhancement`; the comment was just misleading.

Reply:
I have now updated this comment again.

Reply:
> The fact is that an Enhancement request *is* a Status request.

Not so. Otherwise it wouldn't have to requeue a Status request when `queue_status_retrieval` is true. It doesn't directly update the mined height in the case where `d_tx.mined_height().is_none()`.

Reply:
> Otherwise it wouldn't have to requeue a Status request when queue_status_retrieval is true.

The re-queueing is so that the wallet continues to monitor for when the transaction is eventually mined. A relevant case here is that `store_decrypted_tx` was called with a transaction that was read from the mempool; we have complete transaction data, but we don't know yet whether the transaction will be mined, so we stick the status request back on the queue if we have no other way of detecting that it becomes mined.